modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/util/TypeUtils.java

```diff
@@ -348,16 +363,41 @@ else if (val instanceof Number && storageType != val.getClass()) {
             return val;
     }
 
+    /** Converts temporal objects to long.
+     *
+     * @param ctx Data context.
+     * @param val Temporal value.
+     * @return Millis value.
+     */
+    private static long toLong(DataContext ctx, Object val) {
+        if (val instanceof LocalDateTime)
+            return SqlFunctions.toLong(Timestamp.valueOf((LocalDateTime)val), DataContext.Variable.TIME_ZONE.get(ctx));
+
+        if (val instanceof LocalDate)
+            return SqlFunctions.toLong(DateValueUtils.convertToSqlDate((LocalDate)val), DataContext.Variable.TIME_ZONE.get(ctx));
+
+        if (val instanceof LocalTime)
+            return SqlFunctions.toLong(DateValueUtils.convertToSqlTime((LocalTime)val), DataContext.Variable.TIME_ZONE.get(ctx));
+
+        return SqlFunctions.toLong((java.util.Date)val, DataContext.Variable.TIME_ZONE.get(ctx));
+    }
+
     /** */
     public static Object fromInternal(DataContext ctx, Object val, Type storageType) {
         if (val == null)
             return null;
         else if (storageType == java.sql.Date.class && val instanceof Integer)
             return new java.sql.Date(fromLocalTs(ctx, (Integer)val * DateTimeUtils.MILLIS_PER_DAY));
+        else if (storageType == LocalDate.class && val instanceof Integer)
+            return Instant.ofEpochMilli((Integer)val * DateTimeUtils.MILLIS_PER_DAY).atZone(ZoneOffset.UTC).toLocalDate();
```

Comment:
As far as I remember `ofEpochMilli` return different values than `Date` constructor for dates before `GregorianCalendar#gregorianCutoverYearJulian`, can you please check it?